### PR TITLE
Scripts to download HF dataset in arrayrecord or parquet

### PR DESCRIPTION
**Pull request by aireenmei**
# Description

Add scripts for downloading Hugging Face dataset in ArrayRecord or Parquet formats.
See the docstring in the script for details.

# Tests
Use the scripts to download the OptimalScale/ClimbMix dataset
* Download as ArrayRecord
```
python download_hf_dataset_as_arrayrecord.py \
    --dataset OptimalScale/ClimbMix \
    --output <path> \
    --workers 16
```
Log output: `Done! Wrote 553,315,056 records to 1665 files in 16245.8s`
Files are in `gs://maxtext-dataset/array-record/climbmix/*.arrayrecord`

* Download as Parquet
```
python download_hf_dataset_as_parquet.py \
    --dataset OptimalScale/ClimbMix \
    --num-files 2048 \
    --row-group-size 50000 \
    --output <path> \
    --workers 16 \
    --split train
```
log output: `Done! Wrote 553,315,056 rows in 12535.5s`
Files are in `gs://maxtext-dataset/hf/climbmix/*.parquet`

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: tools/data_generation/download_hf_dataset_as_arrayrecord.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+
+
+def rename_to_convention(output_dir, all_filenames, prefix):
+  """Rename worker temp files to a contiguous naming scheme."""
+  # Filter for existing, non-empty files first
+  print("\nVerifying downloaded files...")
+  existing_files = []
+  for fname in all_filenames:
+    fpath = os.path.join(output_dir, fname)
+    if os.path.exists(fpath) and os.path.getsize(fpath) > 0:
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tools/data_generation/download_hf_dataset_as_arrayrecord.py:341]: updated width to a min of 5

File: tools/data_generation/download_hf_dataset_as_parquet.py [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+          "token": args.token,
+          "output": args.output,
+          "row_group_size": args.row_group_size,
+          "compression": args.compression,
+          "jsonl_bytes_per_file": jsonl_bytes_per_file,
+      }
+      for i in range(num_workers)
+  ]
+
+  print(f"Plan: {total_bytes/1e9:.1f} GB -> ~{args.num_files} files | Workers: {num_workers}\n")
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tools/data_generation/download_hf_dataset_as_parquet.py:392]: updated width to a min of 5

File: tools/data_generation/download_hf_dataset_as_parquet.py [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+        ckpt = json.load(f)
+      ds.load_state_dict(ckpt["ds_state"])
+      return ds, ckpt
+    return ds, {"local_file_idx": 0, "total_jsonl_bytes": 0, "total_written": 0, "filenames": []}
+
+  ds, state = load_and_restore()
+  local_file_idx, total_jsonl_bytes, total_written, filenames = (
+      state[k] for k in ["local_file_idx", "total_jsonl_bytes", "total_written", "filenames"]
+  )
+  fs, base_path = get_filesystem_and_path(output_path)
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tools/data_generation/download_hf_dataset_as_parquet.py:253]: fixed. Will use local path for checkpoints and user provides gs:// as output path

File: tools/data_generation/download_hf_dataset_as_parquet.py [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+        "worker_id": worker_id,
+        "local_file_idx": local_file_idx,
+        "total_jsonl_bytes": total_jsonl_bytes,
+        "total_written": total_written,
+        "filenames": filenames,
+        "ds_state": ds.state_dict(),
+    }
+    with open(checkpoint_path, "w", encoding="utf-8") as f:
+      json.dump(ckpt, f)
+
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tools/data_generation/download_hf_dataset_as_parquet.py:282]: Switched to using a local path for checkpoints and user provides gs:// as output path

File: tools/data_generation/download_hf_dataset_as_parquet.py [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+  rename_to_hf_convention(
+      fs, base_path, all_filenames, args.name_prefix or args.dataset.split("/")[-1].lower().replace(" ", "-")
+  )
+
+  # Clean up checkpoint files
+  for i in range(num_workers):
+    path = os.path.join(os.path.abspath(args.output), f".checkpoint-worker-{i:04d}.json")
+    if os.path.exists(path):
+      os.remove(path)
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tools/data_generation/download_hf_dataset_as_parquet.py:435]: Switched to using a local path for checkpoints and user provides gs:// as output path

File: tools/data_generation/download_hf_dataset_as_arrayrecord.py [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+        record = serialize_example(example)
+        record_size = len(record)
+        writer.write(record)
+        file_bytes += record_size
+        total_bytes += record_size
+        total_written += 1
+
+        if file_bytes >= adjusted_file_size_bytes:
+          if writer:
+            writer.close()
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tools/data_generation/download_hf_dataset_as_arrayrecord.py]: When a connection error occurs, the checkpoint is loaded on line 303

File: tools/data_generation/download_hf_dataset_as_parquet.py [hunk 7]
```diff
@@ -0,0 +1,425 @@
...
+  width = len(str(total_files))
+
+  for idx, old_name in enumerate(existing_files):
+    new_name = f"{prefix}-{idx:0{width}d}-of-{total_files:0{width}d}.parquet"
+    if old_name != new_name:
+      fs.move(base_path + old_name, base_path + new_name)
+
+  if total_files > 0:
+    print(f"  Renamed to: {prefix}-00000-of-{total_files:0{width}d}.parquet ...")
+
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply tools/data_generation/download_hf_dataset_as_parquet.py]: When a connection error occur, checkpoint is loaded on line 281

File: tools/data_generation/download_hf_dataset_as_arrayrecord.py [hunk 8]
```diff
@@ -0,0 +1,422 @@
+#!/usr/bin/env python3
```
Comment (gagika) [inline tools/data_generation/download_hf_dataset_as_arrayrecord.py]: do they work only for text only dataset or multi-modal too?
if it's text only can you add a comment about it?

**Review by gagika — APPROVED** [review]
thanks

File: tools/data_generation/download_hf_dataset_as_arrayrecord.py (hunk 8 above)
Comment (aireenmei) [reply tools/data_generation/download_hf_dataset_as_arrayrecord.py]: Good point, text only now. Added in description

Comment (aireenmei) [reply tools/data_generation/download_hf_dataset_as_arrayrecord.py]: Good point, text only now. Added in description.
